src/core/io_queue.cc

```diff
@@ -1055,7 +1059,26 @@ void io_queue::submit_request(queued_io_request_completion* desc, internal::io_r
     _queued_requests--;
     _requests_executing++;
     _requests_dispatched++;
-    _sink.submit(desc, std::move(req));
+
+    if (req.opcode() != internal::io_request::operation::discard) {
+        _sink.submit(desc, std::move(req));
+    } else {
+        submit_blocks_discarding(desc, std::move(req));
+    }
+}
+
+void io_queue::submit_blocks_discarding(queued_io_request_completion* desc, internal::io_request req) noexcept {
+    auto& discard_req = req.as<internal::io_request::operation::discard>();
+
+    // Caller synchronizes using the future returned from `desc->get_future()`.
+    // The attached continuation resolves that future when `punch_hole()` finishes.
+    (void) engine().punch_hole(discard_req.fd, discard_req.offset, discard_req.length).then_wrapped([desc, len = discard_req.length] (future<> f) {
```

Comment:
nit: There's neat `future::forward_to(promise<>)` method that can be used to link punch_hole's future with desc's promise directly

Reply:
I tried to use it, but the following two issues occurred.

Firstly, the type of the future returned from `punch_hole()` is different than the type of future produced by the promise from `desc`.
```
/repo/scylladb/seastar/seastar/include/seastar/core/future.hh:1610:34: note:   no known conversion for argument 1 from ‘seastar::promise<long unsigned int>’ to ‘seastar::promise<void>&&’
```

Secondly, `desc->set_exception()` and `desc->complete()` are part of reasoning about the lifetime of `queued_io_request_completion`. They delete `desc` object -- the memory is freed. Moreover, they update `priority_class_data` and `io_queue`.

Reply:
> Firstly, the type of the future returned from punch_hole() is different than the type of future produced by the promise from desc.

There's future::ignore_result() that converts future<T> into future<>